src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/PartitionEliminationRewriter.cs

```diff
@@ -45,28 +45,25 @@ internal class PartitionEliminationRewriter : SqlExpressionRewriterWithInitialCo
             _resourceTypeSearchParameter = searchParameterDefinitionManagerResolver.Invoke().GetSearchParameter(KnownResourceTypes.Resource, SearchParameterNames.ResourceType);
         }
 
-        private SearchParameterExpression AllTypesExpression
+        private SearchParameterExpression GetAllTypesExpression()
         {
-            get
+            if (_allTypesExpression != null)
             {
-                if (_allTypesExpression != null)
-                {
-                    return _allTypesExpression;
-                }
+                return _allTypesExpression;
+            }
 
-                var typeExpressions = new Expression[_model.ResourceTypeIdRange.highestId - _model.ResourceTypeIdRange.lowestId + 1];
+            object[] resourceTypes = new object[_model.ResourceTypeIdRange.highestId - _model.ResourceTypeIdRange.lowestId + 1];
 
-                for (short i = 0, typeId = _model.ResourceTypeIdRange.lowestId; typeId <= _model.ResourceTypeIdRange.highestId; typeId++, i++)
-                {
-                    typeExpressions[i] = Expression.StringEquals(FieldName.TokenCode, null, _model.GetResourceTypeName(typeId), false);
-                }
+            for (short i = 0, typeId = _model.ResourceTypeIdRange.lowestId; typeId <= _model.ResourceTypeIdRange.highestId; typeId++, i++)
+            {
+                resourceTypes[i] = _model.GetResourceTypeName(typeId);
+            }
 
-                _allTypesExpression = Expression.SearchParameter(
-                    _resourceTypeSearchParameter,
-                    Expression.Or(typeExpressions));
+            _allTypesExpression = Expression.SearchParameter(
+               _resourceTypeSearchParameter,
+               Expression.In(FieldName.TokenCode, null, resourceTypes));
```

Comment:
Should we have made `Expression.In<T>`? Seeing the above in a code review makes me think the object array will box all values where perhaps we could easily avoid it?